libs/s25main/PickedMovableObject.cpp

```diff
@@ -0,0 +1,172 @@
+// Copyright (C) 2005 - 2023 Settlers Freaks (sf-team at siedler25.org)
+//
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "PickedMovableObject.h"
+#include "desktops/dskGameInterface.h"
+#include "drivers/VideoDriverWrapper.h"
+#include "world/GameWorldBase.h"
+#include "world/GameWorldView.h"
+#include "nodeObjs/noMovable.h"
+#include <cmath>
+
+namespace {
+constexpr unsigned PickRadius = 2;
+constexpr unsigned TrackRadius = 5;
+constexpr unsigned long ExpireIn = 5000;
+constexpr auto ExpireNever = static_cast<unsigned long>(-1);
+
+// Make (ab-)use of CheckPointsInRadius() easier
+constexpr bool CheckPointsBreak = true;
+constexpr bool CheckPointsContinue = false;
+} // namespace
+
+PickedMovableObject PickedMovableObject::Pick(const GameWorldView& gwv, MapPoint mapPt, DrawPoint drawPt, bool expire)
+{
+    // DEBUG REMOVE BEFORE MERGE
+    unsigned i = 1;
+
+    const auto offset = gwv.GetOffset();
+    const auto center = gwv.GetSize() / 2.f;
+    const auto zoomFactor = gwv.GetZoomFactor();
+    const auto& world = gwv.GetWorld();
+    const auto worldSize = world.GetSize() * DrawPoint(TR_W, TR_H);
+    auto minDistance = std::numeric_limits<float>::max();
+
+    PickedMovableObject pmo;
+
+    world.CheckPointsInRadius(
+      mapPt, PickRadius,
+      [&](MapPoint curPt, unsigned) {
+          if(gwv.GetViewer().GetVisibility(curPt) != Visibility::Visible)
+              return CheckPointsContinue;
+
+          DrawPoint curDrawPt = world.GetNodePos(curPt);
+          if(curDrawPt.x < offset.x)
+              curDrawPt.x += worldSize.x;
+          if(curDrawPt.y < offset.y)
+              curDrawPt.y += worldSize.y;
+          curDrawPt -= offset;
+          for(const noBase& obj : world.GetFigures(curPt))
+          {
+              const auto* movable = dynamic_cast<const noMovable*>(&obj);
+              if(!movable)
+                  continue;
+
+              DrawPoint objDrawPt = curDrawPt;
+              if(movable->IsMoving())
+                  objDrawPt += movable->CalcWalkingRelative();
+              objDrawPt = DrawPoint((objDrawPt - center) * zoomFactor + center);
+
+              // DEBUG REMOVE BEFORE MERGE
+              dskGameInterface::SetDebugPoint(i++, objDrawPt, 8, 4, MakeColor(255, 255, 0, 255));
+
+              auto diff = Point<float>(objDrawPt - drawPt);
+              float distance = std::sqrt(diff.x * diff.x + diff.y * diff.y);
```

Comment:
```suggestion
              const auto diff = Point<float>(objDrawPt - drawPt);
              const float distance = std::sqrt(diff.x * diff.x + diff.y * diff.y);
```

Might even be worth to have a function for that ([cartesian] distance between 2 points)

Reply:
I did use `boost::math::hypot()` once. Should I use that again or add my own?

Reply:
I was thinking about a `U distance(Point, Point)` function or even a member function so `const auto distance = objDrawPt.distance(drawPt)` could be used as we use this pattern in at least 1 other place.

There is even `std::hypot` to be used in the implementation.

Reply:
Unfortunately the Apple cross-compiler doesn't support anything newer. It was hard to get C++14 already which I think is a good one with newer ones just icing on the cake.

`std::exchange` is C++14: https://en.cppreference.com/w/cpp/utility/exchange

> std::hypot() was added in C++11.

I looked up the Boost docs and found a comment referring to the standard, so quickly checked when it was added. Otherwise wouldn't have known either.